From 3f39c0c87dfe0c58d4421b23a60735d515021fc7 Mon Sep 17 00:00:00 2001
From: Jeff King <peff@peff.net>
Date: Wed, 4 Jan 2012 09:28:23 -0500
Subject: [PATCH] receive-pack: respect receive.advertiseAlternates config

Usually receive-pack advertises ref tips from alternates
repositories so that clients can sometimes avoid sending
objects that are already upstream.

However, if you have a very large alternates network, then
the number of ".have" refs can get cumbersome, and you spend
more time advertising refs that the client doesn't care
about than you are saving in the optimization.

This patch adds a config variable to drop .have lines
entirely.

An alternative approach would be to restrict the .have lines
to a smaller portion of the namespace that is known to be
"interesting" to most clients. This patch is much simpler;
if the loss of the .have optimization turns out to be too
much, we can try something more complex.

Signed-off-by: Jeff King <peff@peff.net>
---
 builtin/receive-pack.c |   16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/builtin/receive-pack.c b/builtin/receive-pack.c
index 0afb8b2..3f405df 100644
--- a/builtin/receive-pack.c
+++ b/builtin/receive-pack.c
@@ -40,6 +40,7 @@ static int auto_gc = 1;
 static const char *head_name;
 static void *head_name_to_free;
 static int sent_capabilities;
+static int advertise_alternates = 1;
 
 static enum deny_action parse_deny_action(const char *var, const char *value)
 {
@@ -113,6 +114,11 @@ static int receive_pack_config(const char *var, const char *value, void *cb)
 		return 0;
 	}
 
+	if (strcmp(var, "receive.advertisealternates") == 0) {
+		advertise_alternates = git_config_bool(var, value);
+		return 0;
+	}
+
 	return git_default_config(var, value, cb);
 }
 
@@ -158,10 +164,12 @@ static void collect_one_alternate_ref(const struct ref *ref, void *data)
 
 static void write_head_info(void)
 {
-	struct sha1_array sa = SHA1_ARRAY_INIT;
-	for_each_alternate_ref(collect_one_alternate_ref, &sa);
-	sha1_array_for_each_unique(&sa, show_one_alternate_sha1, NULL);
-	sha1_array_clear(&sa);
+	if (advertise_alternates) {
+		struct sha1_array sa = SHA1_ARRAY_INIT;
+		for_each_alternate_ref(collect_one_alternate_ref, &sa);
+		sha1_array_for_each_unique(&sa, show_one_alternate_sha1, NULL);
+		sha1_array_clear(&sa);
+	}
 	for_each_ref(show_ref_cb, NULL);
 	if (!sent_capabilities)
 		show_ref("capabilities^{}", null_sha1);
-- 
1.7.9.6.6.g6b3b56

